File: app/src/main/java/com/commit451/gitlab/adapter/AddIssueLabelAdapter.java
package com.commit451.gitlab.adapter;

import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.commit451.gitlab.model.api.Label;
import com.commit451.gitlab.viewHolder.AddLabelViewHolder;

import java.util.ArrayList;
import java.util.Collection;

/**
 * So many labels
 */
public class AddIssueLabelAdapter extends RecyclerView.Adapter<AddLabelViewHolder> {

    private ArrayList<Label> values;
    private Listener listener;

    public AddIssueLabelAdapter(Listener listener) {
        values = new ArrayList<>();
        this.listener = listener;
    }

    public void setLabels(Collection<Label> labels) {
        values.clear();
        addLabels(labels);
    }

    public void addLabels(Collection<Label> labels) {
        if (labels != null) {
            values.addAll(labels);
        }
        notifyDataSetChanged();
    }

    public void addLabel(Label label) {
        values.add(label);
        notifyItemInserted(values.size()-1);
    }

    public void removeLabel(Label label) {
        int indexOf = values.indexOf(label);
        values.remove(indexOf);
        notifyItemRemoved(indexOf);
    }

    @Override
    public AddLabelViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final AddLabelViewHolder holder = AddLabelViewHolder.inflate(parent);
        holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                Label label = getEntry(holder.getAdapterPosition());
                listener.onLabelLongClicked(label);
                return true;
            }
        });
        return holder;
    }

    @Override
    public void onBindViewHolder(final AddLabelViewHolder holder, int position) {
        holder.bind(getEntry(position));
    }

    @Override
    public int getItemCount() {
        return values.size();
    }

    private Label getEntry(int position) {
        return values.get(position);
    }

    public boolean containsLabel(Label label) {
        return values.contains(label);
    }

    @Nullable
    public String getCommaSeperatedStringOfLabels() {
        if (values.isEmpty()) {
            return null;
        }
        String labels = "";
        for (Label label : values) {
            labels = labels + label.getName() + ",";
        }
        //Remove last ","
        labels = labels.substring(0, labels.length()-1);
        return labels;
    }

    public interface Listener {
        void onLabelLongClicked(Label label);
    }
}

File: app/src/main/java/com/commit451/gitlab/adapter/AddIssueLabelAdapter.kt
package com.commit451.gitlab.adapter

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.commit451.adapterlayout.AdapterLayout
import com.commit451.gitlab.model.api.Label
import com.commit451.gitlab.viewHolder.AddLabelViewHolder
import java.util.*

/**
 * So many labels
 */
class AddIssueLabelAdapter(private val listener: AddIssueLabelAdapter.Listener) : RecyclerView.Adapter<AddLabelViewHolder>() {

    private val values: ArrayList<Label> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AddLabelViewHolder {
        val holder = AddLabelViewHolder.inflate(parent)
        holder.itemView.setOnClickListener {
            val position = AdapterLayout.getAdapterPosition(holder)
            val label = getEntry(position)
            listener.onLabelClicked(label)
        }
        return holder
    }

    override fun onBindViewHolder(holder: AddLabelViewHolder, position: Int) {
        holder.bind(getEntry(position))
    }

    override fun getItemCount(): Int {
        return values.size
    }

    fun setLabels(labels: Collection<Label>) {
        values.clear()
        addLabels(labels)
    }

    fun addLabels(labels: Collection<Label>?) {
        if (labels != null) {
            values.addAll(labels)
        }
        notifyDataSetChanged()
    }

    fun addLabel(label: Label) {
        values.add(label)
        notifyItemInserted(values.size - 1)
    }

    fun removeLabel(label: Label) {
        val indexOf = values.indexOf(label)
        values.removeAt(indexOf)
        notifyItemRemoved(indexOf)
    }

    fun containsLabel(label: Label): Boolean {
        return values.contains(label)
    }

    private fun getEntry(position: Int): Label {
        return values[position]
    }

    //Remove last ","
    fun getCommaSeperatedStringOfLabels(): String? {
        if (values.isEmpty()) {
            return null
        }
        var labels = ""
        for (label in values) {
            labels = labels + label.name + ","
        }
        //Remove last ","
        labels = labels.substring(0, labels.length - 1)
        return labels
    }

    interface Listener {
        fun onLabelClicked(label: Label)
    }
}
